File: packages/apple/Sources/Runtime/ChainExecutor.swift
import Foundation

// MARK: - Execution Record

/// Record of a chain execution for history and debugging
public struct ChainExecutionRecord: Sendable {
    public let chainName: String
    public let input: String
    public let output: String?
    public let processingTimeMs: Int
    public let success: Bool
    public let attempt: Int
    public let timestamp: Double
}

// MARK: - Chain Executor

/// Chain execution engine with instrumentation, retries, and history tracking.
///
/// ```swift
/// let executor = ChainExecutor(maxRetries: 1)
/// let output = try await executor.execute(summarizeChain, input: ChainInput(text: article))
/// print(executor.getHistory())
/// ```
@available(iOS 15.0, macOS 14.0, *)
public final class ChainExecutor: @unchecked Sendable {
    private var history: [ChainExecutionRecord] = []
    private let maxRetries: Int

    public init(maxRetries: Int = 1) {
        self.maxRetries = maxRetries
    }

    /// Execute a chain with timing, retries, and history tracking
    public func execute(_ chain: any Chain, input: ChainInput) async throws -> ChainOutput {
        let startTime = Date().timeIntervalSince1970
        var lastError: Error?

        for attempt in 0...maxRetries {
            do {
                let output = try await chain.invoke(input)
                let endTime = Date().timeIntervalSince1970
                history.append(ChainExecutionRecord(
                    chainName: chain.name,
                    input: input.text,
                    output: output.text,
                    processingTimeMs: Int((endTime - startTime) * 1000),
                    success: true,
                    attempt: attempt + 1,
                    timestamp: startTime
                ))
                return output
            } catch {
                lastError = error
                if attempt < maxRetries {
                    try? await Task.sleep(nanoseconds: 100_000_000)
                }
            }
        }

        let endTime = Date().timeIntervalSince1970
        history.append(ChainExecutionRecord(
            chainName: chain.name,
            input: input.text,
            output: nil,
            processingTimeMs: Int((endTime - startTime) * 1000),
            success: false,
            attempt: maxRetries + 1,
            timestamp: startTime
        ))
        throw lastError ?? LocanaraError.executionFailed("Chain execution failed")
    }

    /// Get execution history
    public func getHistory() -> [ChainExecutionRecord] { history }

    /// Clear execution history
    public func clearHistory() { history.removeAll() }
}

File: packages/android/locanara/src/main/kotlin/com/locanara/runtime/ChainExecutor.kt
package com.locanara.runtime

import com.locanara.composable.Chain
import com.locanara.core.ChainInput
import com.locanara.core.ChainOutput
import kotlinx.coroutines.delay

/**
 * Record of a chain execution for history and debugging
 */
data class ChainExecutionRecord(
    val chainName: String,
    val input: String,
    val output: String?,
    val processingTimeMs: Int,
    val success: Boolean,
    val attempt: Int,
    val timestamp: Double
)

/**
 * Chain execution engine with instrumentation, retries, and history tracking.
 *
 * ```kotlin
 * val executor = ChainExecutor(maxRetries = 1)
 * val output = executor.execute(summarizeChain, ChainInput(text = article))
 * println(executor.getHistory())
 * ```
 */
class ChainExecutor(
    private val maxRetries: Int = 1
) {
    private val history = mutableListOf<ChainExecutionRecord>()

    /** Execute a chain with timing, retries, and history tracking */
    suspend fun execute(chain: Chain, input: ChainInput): ChainOutput {
        val startTime = System.currentTimeMillis()
        var lastError: Exception? = null

        for (attempt in 0..maxRetries) {
            try {
                val output = chain.invoke(input)
                val endTime = System.currentTimeMillis()
                history.add(
                    ChainExecutionRecord(
                        chainName = chain.name,
                        input = input.text,
                        output = output.text,
                        processingTimeMs = (endTime - startTime).toInt(),
                        success = true,
                        attempt = attempt + 1,
                        timestamp = startTime.toDouble()
                    )
                )
                return output
            } catch (e: Exception) {
                lastError = e
                if (attempt < maxRetries) {
                    delay(100)
                }
            }
        }

        val endTime = System.currentTimeMillis()
        history.add(
            ChainExecutionRecord(
                chainName = chain.name,
                input = input.text,
                output = null,
                processingTimeMs = (endTime - startTime).toInt(),
                success = false,
                attempt = maxRetries + 1,
                timestamp = startTime.toDouble()
            )
        )
        throw lastError ?: IllegalStateException("Chain execution failed")
    }

    /** Get execution history */
    fun getHistory(): List<ChainExecutionRecord> = history.toList()

    /** Clear execution history */
    fun clearHistory() { history.clear() }
}
